Initial P1Events page scroll depth tacking

B77Mills opened the pull request:
New P1 Event  tracking components help track page scroll depth.  The intent is to fire p1 events based on how far the user scrolls down the page.  This will use a querySelecort to determine the end of the target page and calculate scroll depth accordingly. 

The example inclusions for the website look something like this at moment  with default events firing at25, 50, 75 & 90%.

### NativeX story:
Here is a [link](http://www-p1-sandbox.dev.parameter1.com:51269/sponsored/parameter1/another-story/62b37e698a17bc0001015d90?preview=true)

**Template:**
```marko.js
<marko-web-p1-events-track-native-story-scroll-depth story=story  selector=".document-container" />
```

**EventsObj:**

```js
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "25 percent",
            "ent": {
                "id": "native-x.example.story*6309207d5a33e600013c5b83",
                "name": "The ABCs of VIP: How the VIP Experience is Evolving at Sports Venues Across the Country",
                "props": {
                    "type": "story",
                    "published": 1659330000000
                },
                "refs": {
                    "advertiser": {
                        "id": "native-x.example.advertiser*5fcfb9422205c1026a1ebf3e",
                        "name": "Parameter1",
                        "refs": {}
                    }
                }
            }
        }
    ]
},
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "90 percent",
            "ent": {
                "id": "native-x.example.story*6309207d5a33e600013c5b83",
                "name": "The ABCs of VIP: How the VIP Experience is Evolving at Sports Venues Across the Country",
                "props": {
                    "type": "story",
                    "published": 1659330000000
                },
                "refs": {
                    "advertiser": {
                        "id": "native-x.example.advertiser*5fcfb9422205c1026a1ebf3e",
                        "name": "Parameter1",
                        "refs": {}
                    }
                }
            }
        }
    ]
}
```

### Content Page:
This is a good example of [content pages](http://www-p1-sandbox.dev.parameter1.com:51269/home/article/22288696/how-to-manage-and-reduce-risk-at-large-events)

**Template:**
```marko.js
<marko-web-p1-events-track-content-scroll-depth content=content selector=".document-container .page .row .content-page-body" />
```

**EventsObj:**

```js
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "25 percent",
            "ent": {
                "id": "base.p1-sandbox.content-article*22288696",
                "name": "How to Manage and Reduce Risk at Large Events",
                "props": {
                    "type": "article",
                    "published": 1655837711000
                },
                "refs": {
                    "primarySection": {
                        "id": "base.p1-sandbox.website-section*64017",
                        "name": "Home",
                        "props": {
                            "alias": "home"
                        },
                        "refs": {}
                    },
                    "company": null,
                    "authors": [],
                    "createdBy": {
                        "id": "base.p1-sandbox.user*5eea498723013d2b008b4571",
                        "name": "jworden",
                        "props": {
                            "firstName": "Josh",
                            "lastName": "Worden"
                        },
                        "refs": {}
                    }
                }
            }
        }
    ]
},
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "90 percent",
            "ent": {
                "id": "base.p1-sandbox.content-article*22288696",
                "name": "How to Manage and Reduce Risk at Large Events",
                "props": {
                    "type": "article",
                    "published": 1655837711000
                },
                "refs": {
                    "primarySection": {
                        "id": "base.p1-sandbox.website-section*64017",
                        "name": "Home",
                        "props": {
                            "alias": "home"
                        },
                        "refs": {}
                    },
                    "company": null,
                    "authors": [],
                    "createdBy": {
                        "id": "base.p1-sandbox.user*5eea498723013d2b008b4571",
                        "name": "jworden",
                        "props": {
                            "firstName": "Josh",
                            "lastName": "Worden"
                        },
                        "refs": {}
                    }
                }
            }
        }
    ]
}
```

### Website section:

**Template:**
```marko.js
<marko-web-p1-events-track-website-section-scroll-depth section=section selector=".document-container" />
```
**EventObj:**
```js
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Website Section",
            "lab": "25 percent",
            "ent": {
                "id": "base.p1-sandbox.website-section*83028",
                "refs": {}
            }
        }
    ]
}{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Website Section",
            "lab": "25 percent",
            "ent": {
                "id": "base.p1-sandbox.website-section*83028",
                "refs": {}
            }
        }
    ]
},
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Website Section",
            "lab": "90 percent",
            "ent": {
                "id": "base.p1-sandbox.website-section*83028",
                "refs": {}
            }
        }
    ]
}
```

### Static & non-standard page calls
**Template:**
```marko.js 
import ns from "@parameter1/base-cms-marko-web-p1-events/utils/create-namespace";
<marko-web-p1-events-track-static-scroll-depth
  entity={
    id: "home",
      ns: ns("website-section"),
        name: "home",
          alias: "home",
  }
  category="Static"
selector = ".document-container"
  />
```
**EventsObj:**
```js
{
  slug: "p1",
  realm: "5ed294c6c13a4626008b4568",
  env: "development",
  host: "www-p1-sandbox.dev.parameter1.com",
  vis: "FmBWTh5gwTurR0GYdT9It",
  idt: "omeda.hcl.customer*5234E7727901C9J~encrypted",
  events: [
    {
      act: "Scroll",
      cat: "Static",
      lab: "25 percent",
      ent: {
        id: "base.p1-sandbox.website-section*home",
        name: "home",
        refs: {}
      }
    }
  ]
  },
  {
    slug: "p1",
    realm: "5ed294c6c13a4626008b4568",
    env: "development",
    host: "www-p1-sandbox.dev.parameter1.com",
    vis: "FmBWTh5gwTurR0GYdT9It",
    idt: "omeda.hcl.customer*5234E7727901C9J~encrypted",
    events: [
      {
        act: "Scroll",
        cat: "Static",
        lab: "90 percent",
        ent: {
          id: "base.p1-sandbox.website-section*home",
          name: "home",
          refs: { }
        }
      }
    ]
  }
];

```

B77Mills committed:
Initial setup of inbody scroll depth tracking.

This is the initial attempt to track scroll depth on content body pages.  At the moment it will take a body target & trigger console.warn for a p1event track event.  This is currently set to fire at 0%, 25%, 50%, 75% 90%(full/100%) of page view scroll depths.

B77Mills committed:
Fix selector prop setting and use

B77Mills committed:
Adjust how page height and target are set

Adjust querySelector to be querySelectorAll in order to account for multple sibling div to be use to generate the start/end & % scrolled on page.

@todo Determine if there is a better way to wrap or collect & determin where a page starts/ends and how to calculate the % scrolled though.

B77Mills committed:
update tracker to only use single selector

This will allow for us to correctly calculate the % of content body read.

B77Mills committed:
Remove calc of top and us 0 for top of page

Assume top of page is included in content pages.

B77Mills committed:
fix missing ;

B77Mills committed:
Add nonContent page support with pageViewDepth

Update component to take an object vs just content with will still requiree a id and type, with chouls be something linke 12345678 News, ore 12345, Section

B77Mills committed:
call page view depth component on homepage & section page

B77Mills committed:
Remove $nextTick() on setting of container metrics

Original thought was to wait to $nextTick in hope to hone cause issue with lazyloading of inbody images, but I dont think that will be an issue.

B77Mills committed:
remove the default 0.1/pageload & up fullView to 1

Remove the 0.1 page loaded intial sroll page depth trigger and up the fullViewed depth to 100% vs 90%

B77Mills committed:
Update action name generation

Update the action string formater to be `${Number(d) * 100} percent` to better match what we had been  sendin to other metric trackers like UA/GA

B77Mills committed:
Replace console.warn with p1event track event

B77Mills committed:
update Id prop to be string to support sectionAlias

This will help support things like dynamic pages and homepages or things like /top-companies

B77Mills committed:
update eventEntity util to be generic base entity

this will generat the `${type}-${id}` vs `content-${type}-$id}`

brandonbk commented on packages/marko-web-p1-events/components/marko.json:
```diff
@@ -37,6 +37,24 @@
     "@action": "string",
     "@category": "string"
   },
+  "<marko-web-p1-events-track-page-view-depth>": {
```
Rename to `marko-web-p1-events-track-scroll-depth`

brandonbk commented on packages/marko-web-p1-events/components/track-page-view-depth.marko:
```diff
@@ -0,0 +1,14 @@
+$ const obj = getAsObject(input, "obj");
+$ const { selector } = input;
+
+<marko-web-browser-component
```
Rename to `P1EventsTrackScrollDepth` and rename file

brandonbk commented on services/example-website/server/templates/content.marko:
```diff
@@ -23,6 +23,7 @@ $ const { recaptcha } = out.global;
         <div class="col-md-8">
           <div class="content-page-body">
             <theme-page-contents|{ blockName }| attrs={ "data-gallery-id": id }>
+            <marko-web-p1-events-track-page-view-depth obj={id: content.id, type: content.type } />
```
Rename to `marko-web-p1-events-track-scroll-depth`

brandonbk commented on services/example-website/server/templates/index.marko:
```diff
@@ -4,6 +4,7 @@
   description="This is a test page."
 >
   <@page>
+    <marko-web-p1-events-track-page-view-depth obj={ id: 'home', type: 'section' } selector=".document-container" />
```
Rename to `marko-web-p1-events-track-scroll-depth`

brandonbk commented on services/example-website/server/templates/section.marko:
```diff
@@ -3,7 +3,9 @@ $ const { recaptcha } = out.global;
   <@page>
+
     <marko-web-resolve-page|{ data: section, resolved }| node=pageNode>
+    <marko-web-p1-events-track-page-view-depth obj={ id: section.id, type: 'Section' } selector=".document-container" />
```
Rename to `marko-web-p1-events-track-scroll-depth`

brandonbk commented on packages/marko-web-p1-events/components/marko.json:
```diff
@@ -37,6 +37,24 @@
     "@category": "string"
   },
+  "<marko-web-p1-events-track-page-view-depth>": {
+    "template": "./track-page-view-depth.marko",
```
Update reference when file is renamed

brandonbk commented on packages/marko-web-p1-events/browser/track-page-view-depth.vue:
```diff
@@ -0,0 +1,98 @@
+    },
+    targetViewDepths: {
+      type: Array,
+      default: () => ([0.25, 0.50, 0.75]),
```
Add 0.90 to list of default depths

brandonbk commented on packages/marko-web-p1-events/browser/index.js:
```diff
@@ -1,10 +1,12 @@
 export default (Browser) => {
   const { EventBus } = Browser;
   Browser.register('P1EventsTrackContentBodyLinks', TrackContentBodyLinks);
+  Browser.register('P1EventsTrackPageViewDepth', TrackPageViewDepth);
```
Rename to `P1EventsTrackScrollDepth`

brandonbk commented on packages/marko-web-p1-events/browser/track-page-view-depth.vue:
```diff
@@ -0,0 +1,98 @@
+    },
+    selector: {
+      type: String,
+      default: '.document-container .page .row .content-page-body',
```
Remove the default and make the selector required

brandonbk commented on packages/marko-web-p1-events/browser/track-page-view-depth.vue:
```diff
@@ -0,0 +1,98 @@
+      type: Array,
+      default: () => ([0.25, 0.50, 0.75]),
+    },
+    action: {
```
Remove this since it is generated

brandonbk committed:
Update content.marko to add space

B77Mills committed:
Rename to trackScrollDepth

B77Mills committed:
Make selector a required prop

B77Mills committed:
More updates of `view` to `scroll` in props & code

B77Mills committed:
Remove template include

Original intent was to possibly use it as a as top of page is start ref, but top of page now that

B77Mills committed:
Revert removal to <template>

B77Mills committed:
remove use of fullScrollDepth prop

B77Mills committed:
Update action, category prop use

Update Action  to Scroll and set the default value for category to Content for now.  Might make this a required prop.  Also update the payload sent to p1Event to include the lab: `${scrollDepth} percent`

B77Mills committed:
setup individual trackScrollDepth components

Split single component into three individual componetns for section, native-x & content which will handle the setting up entity, categories and other assumed model logic

B77Mills committed:
Add static scroll depth tracker

This will be utilized on pages like home page & top compies pages.

solocommand commented on packages/marko-web-p1-events/components/track-scroll-depth/content.marko:
```diff
@@ -0,0 +1,47 @@
+$ const {
+  content,
+  action,
+  category = 'Content',
```
Don't allow action or category to be customized in the specific components -- content should always be content category and action should always be scroll.

The category should be required on your generic/static component, but action should still be hardcoded to scroll.

B77Mills committed:
Update use of action and category props

Only allow/require that category be set on static scroll components otherwise assume and set action and category

B77Mills committed:
Remove static example

if there is a need to have a site or other specific version of this it can be setup or added on a per situation basis.  Home should be able to use the website section version of the component.

B77Mills commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Inbody Scroll depth tracking](https://trello.com/c/YFoNcGcU/242-inbody-scroll-depth-tracking)
